File: src/waypoint/sources/oem/dell_driverpack.py
"""Dell's per-model driver pack catalog.

https://downloads.dell.com/catalog/DriverPackCatalog.cab — verified by
downloading and inspecting the real file (2026-08-30). Each
`DriverPackage` lists `SupportedSystems/Brand/Model` entries with a
`systemID` attribute (Dell's SMBIOS-reported system ID, the same value
Dell Command | Update and Windows' own `Win32_ComputerSystemProduct.
IdentifyingNumber`-adjacent SMBIOS fields expose) and, unlike the
per-device `CatalogPC.cab`, a real `Cryptography/Hash algorithm="SHA256"`
per package — so `DriverPack.hash_algorithm` here is honestly `"sha256"`,
not a fallback.
"""

from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Callable
from datetime import date
from pathlib import Path

from waypoint.sources.oem.cab import extract_cab
from waypoint.sources.oem.http import download_file
from waypoint.sources.oem.model_pack import DriverPack

DEFAULT_CATALOG_URL = "https://downloads.dell.com/catalog/DriverPackCatalog.cab"


class DellDriverPackSource:
    source_id = "dell_driverpack"

    def __init__(
        self,
        cache_dir: str,
        *,
        catalog_url: str = DEFAULT_CATALOG_URL,
        downloader: Callable[[str, str], None] | None = None,
    ) -> None:
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self._catalog_url = catalog_url
        self._downloader = downloader or (lambda url, dest: download_file(url, dest))
        self._catalog_xml_path = self.cache_dir / "DriverPackCatalog.xml"
        self._index: dict[str, list[DriverPack]] | None = None

    def refresh(self, *, force: bool = False) -> None:
        if force or not self._catalog_xml_path.exists():
            import tempfile

            with tempfile.TemporaryDirectory() as tmp:
                cab_path = Path(tmp) / "DriverPackCatalog.cab"
                self._downloader(self._catalog_url, str(cab_path))
                extracted = extract_cab(cab_path, tmp)
                xml_file = next((p for p in extracted if p.suffix.lower() == ".xml"), None)
                if xml_file is None:
                    raise ValueError(f"No .xml payload found inside {self._catalog_url}")
                xml_file.replace(self._catalog_xml_path)
        self._index = None
        self.load_from_xml(str(self._catalog_xml_path))

    def load_from_xml(self, xml_path: str) -> None:
        # DriverPackCatalog.xml declares a default namespace
        # (xmlns="openmanage/cm/dm") on its root element, unlike CatalogPC.xml
        # (no default namespace) — verified against the real downloaded file.
        # The `{*}tag` wildcard matches the tag regardless of namespace so
        # this parses correctly either way.
        tree = ET.parse(xml_path)
        root = tree.getroot()
        base_location = root.attrib.get("baseLocation", "downloads.dell.com")

        index: dict[str, list[DriverPack]] = {}
        for package in root.findall("{*}DriverPackage"):
            models = package.findall("{*}SupportedSystems/{*}Brand/{*}Model")
            if not models:
                continue

            os_names = [
                (os_el.find("{*}Display").text or "").strip()
                for os_el in package.findall("{*}SupportedOperatingSystems/{*}OperatingSystem")
                if os_el.find("{*}Display") is not None
            ]
            os_label = ", ".join(n for n in os_names if n) or "unknown"

            hash_algo, hash_value = _best_hash(package)
            path = package.attrib.get("path", "")
            url = f"https://{base_location}/{path}"
            release_dt = _parse_dell_release_date(package.attrib.get("dateTime", ""))

            for model in models:
                system_id = model.attrib.get("systemID", "")
                if not system_id:
                    continue
                pack = DriverPack(
                    pack_id=package.attrib.get("releaseID", ""),
                    model_name=model.attrib.get("name", "unknown"),
                    model_key=system_id,
                    os_label=os_label,
                    version=package.attrib.get("dellVersion", "unknown"),
                    release_date=release_dt,
                    url=url,
                    hash_algorithm=hash_algo,
                    hash_value=hash_value,
                    size_bytes=int(package.attrib.get("size", 0) or 0),
                    source_id=self.source_id,
                )
                index.setdefault(system_id.upper(), []).append(pack)

        self._index = index

    def packs_for_model(self, model_key: str) -> list[DriverPack]:
        if self._index is None:
            self.load_from_xml(str(self._catalog_xml_path))
        return list(self._index.get(model_key.upper(), []))  # type: ignore[union-attr]


def _best_hash(package: ET.Element) -> tuple[str, str]:
    """Prefer SHA256, fall back to SHA1, then MD5 — whatever the catalog
    actually published for this specific package (older packages in this
    catalog sometimes only carry weaker hashes)."""
    crypto = package.find("{*}Cryptography")
    if crypto is None:
        return "md5", package.attrib.get("hashMD5", "")
    hashes = {h.attrib.get("algorithm", "").upper(): (h.text or "").strip() for h in crypto.findall("{*}Hash")}
    for algo in ("SHA256", "SHA1", "MD5"):
        if hashes.get(algo):
            return algo.lower(), hashes[algo]
    return "md5", package.attrib.get("hashMD5", "")


def _parse_dell_release_date(value: str) -> date | None:
    from datetime import datetime

    if not value:
        return None
    for fmt in ("%Y-%m-%dT%H:%M:%S", "%Y-%m-%d"):
        try:
            return datetime.strptime(value[:19], fmt).date()  # noqa: DTZ007 -- date-only field, no TZ published
        except ValueError:
            continue
    return None

File: rust/crates/sources/src/oem/dell_driverpack.rs
//! Dell's per-model driver pack catalog. Port of
//! `sources/oem/dell_driverpack.py`.
//!
//! <https://downloads.dell.com/catalog/DriverPackCatalog.cab> — verified
//! by downloading and inspecting the real file (2026-08-30, per the
//! Python source's docstring). Each `DriverPackage` lists
//! `SupportedSystems/Brand/Model` entries with a `systemID` attribute
//! (Dell's SMBIOS-reported system ID) and, unlike the per-device
//! `CatalogPC.cab`, a real `Cryptography/Hash algorithm="SHA256"` per
//! package — so `DriverPack.hash_algorithm` here is honestly `"sha256"`,
//! not a fallback.
//!
//! Same lazy-cache-via-`RefCell` note as `dell_catalog.rs` applies here:
//! `packs_for_model` self-heals by loading from `catalog_xml_path` if
//! nothing has been loaded yet.

use std::cell::RefCell;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

use time::Date;

use crate::oem::cab::extract_cab;
use crate::oem::dell_catalog::decode_xml_bytes;
use crate::oem::http::download_file;
use crate::oem::model_pack::{DriverPack, ModelDriverPackSource};

pub const DEFAULT_CATALOG_URL: &str = "https://downloads.dell.com/catalog/DriverPackCatalog.cab";

type Downloader = dyn Fn(&str, &Path) -> Result<PathBuf, String>;
type Extractor = dyn Fn(&Path, &Path) -> Result<Vec<PathBuf>, String>;

pub struct DellDriverPackSource {
    pub cache_dir: PathBuf,
    catalog_url: String,
    catalog_xml_path: PathBuf,
    downloader: Box<Downloader>,
    extractor: Box<Extractor>,
    index: RefCell<Option<HashMap<String, Vec<DriverPack>>>>,
}

impl DellDriverPackSource {
    pub fn new(cache_dir: impl Into<PathBuf>) -> Result<Self, String> {
        Self::with_catalog_url(cache_dir, DEFAULT_CATALOG_URL)
    }

    pub fn with_catalog_url(
        cache_dir: impl Into<PathBuf>,
        catalog_url: impl Into<String>,
    ) -> Result<Self, String> {
        let cache_dir = cache_dir.into();
        fs::create_dir_all(&cache_dir).map_err(|e| e.to_string())?;
        let catalog_xml_path = cache_dir.join("DriverPackCatalog.xml");
        Ok(Self {
            cache_dir,
            catalog_url: catalog_url.into(),
            catalog_xml_path,
            downloader: Box::new(|url, dest| download_file(url, dest).map(|_| dest.to_path_buf())),
            extractor: Box::new(|cab, dest| extract_cab(cab, dest)),
            index: RefCell::new(None),
        })
    }

    pub fn set_downloader(&mut self, downloader: impl Fn(&str, &Path) -> Result<PathBuf, String> + 'static) {
        self.downloader = Box::new(downloader);
    }

    pub fn set_extractor(&mut self, extractor: impl Fn(&Path, &Path) -> Result<Vec<PathBuf>, String> + 'static) {
        self.extractor = Box::new(extractor);
    }

    pub fn refresh(&self, force: bool) -> Result<(), String> {
        if force || !self.catalog_xml_path.exists() {
            let dir = make_tempdir()?;
            let cab_path = dir.join("DriverPackCatalog.cab");
            (self.downloader)(&self.catalog_url, &cab_path)?;
            let extracted = (self.extractor)(&cab_path, &dir)?;
            let xml_file = extracted
                .iter()
                .find(|p| p.extension().map(|e| e.eq_ignore_ascii_case("xml")).unwrap_or(false))
                .ok_or_else(|| format!("No .xml payload found inside {}", self.catalog_url))?;
            fs::rename(xml_file, &self.catalog_xml_path).map_err(|e| e.to_string())?;
            let _ = fs::remove_dir_all(&dir);
        }
        self.load_from_xml(&self.catalog_xml_path)
    }

    pub fn load_from_xml(&self, xml_path: &Path) -> Result<(), String> {
        let raw = fs::read(xml_path).map_err(|e| e.to_string())?;
        let text = decode_xml_bytes(&raw)?;
        let doc = roxmltree::Document::parse(&text).map_err(|e| e.to_string())?;
        let root = doc.root_element();
        let base_location = root.attribute("baseLocation").unwrap_or("downloads.dell.com");

        let mut index: HashMap<String, Vec<DriverPack>> = HashMap::new();

        for package in root.descendants().filter(|n| n.has_tag_name("DriverPackage")) {
            let models: Vec<_> = package
                .descendants()
                .filter(|n| n.has_tag_name("Model") && n.parent().map(|p| p.has_tag_name("Brand")).unwrap_or(false))
                .collect();
            if models.is_empty() {
                continue;
            }

            let os_names: Vec<String> = package
                .descendants()
                .filter(|n| n.has_tag_name("OperatingSystem"))
                .filter_map(|os_el| {
                    os_el
                        .children()
                        .find(|c| c.has_tag_name("Display"))
                        .and_then(|d| d.text())
                        .map(|t| t.trim().to_string())
                })
                .collect();
            let os_label = if os_names.iter().any(|s| !s.is_empty()) {
                os_names.into_iter().filter(|s| !s.is_empty()).collect::<Vec<_>>().join(", ")
            } else {
                "unknown".to_string()
            };

            let (hash_algo, hash_value) = best_hash(package);
            let path = package.attribute("path").unwrap_or("");
            let url = format!("https://{base_location}/{path}");
            let release_dt = parse_dell_driverpack_date(package.attribute("dateTime").unwrap_or(""));
            let pack_id = package.attribute("releaseID").unwrap_or("").to_string();
            let version = package.attribute("dellVersion").unwrap_or("unknown").to_string();
            let size_bytes: u64 = package.attribute("size").and_then(|s| s.parse().ok()).unwrap_or(0);

            for model in models {
                let system_id = model.attribute("systemID").unwrap_or("");
                if system_id.is_empty() {
                    continue;
                }
                let pack = DriverPack {
                    pack_id: pack_id.clone(),
                    model_name: model.attribute("name").unwrap_or("unknown").to_string(),
                    model_key: system_id.to_string(),
                    os_label: os_label.clone(),
                    version: version.clone(),
                    release_date: release_dt,
                    url: url.clone(),
                    hash_algorithm: hash_algo.clone(),
                    hash_value: hash_value.clone(),
                    size_bytes,
                    source_id: "dell_driverpack".to_string(),
                };
                index.entry(system_id.to_uppercase()).or_default().push(pack);
            }
        }

        *self.index.borrow_mut() = Some(index);
        Ok(())
    }
}

impl ModelDriverPackSource for DellDriverPackSource {
    fn source_id(&self) -> &str {
        "dell_driverpack"
    }

    fn packs_for_model(&self, model_key: &str) -> Result<Vec<DriverPack>, String> {
        if self.index.borrow().is_none() {
            self.load_from_xml(&self.catalog_xml_path)?;
        }
        let index = self.index.borrow();
        Ok(index
            .as_ref()
            .unwrap()
            .get(&model_key.to_uppercase())
            .cloned()
            .unwrap_or_default())
    }
}

/// Prefer SHA256, fall back to SHA1, then MD5 — whatever the catalog
/// actually published for this specific package (older packages in this
/// catalog sometimes only carry weaker hashes).
fn best_hash(package: roxmltree::Node) -> (String, String) {
    let Some(crypto) = package.descendants().find(|n| n.has_tag_name("Cryptography")) else {
        return ("md5".to_string(), package.attribute("hashMD5").unwrap_or("").to_string());
    };
    let mut hashes: HashMap<String, String> = HashMap::new();
    for h in crypto.children().filter(|n| n.has_tag_name("Hash")) {
        let algo = h.attribute("algorithm").unwrap_or("").to_uppercase();
        let value = h.text().unwrap_or("").trim().to_string();
        hashes.insert(algo, value);
    }
    for algo in ["SHA256", "SHA1", "MD5"] {
        if let Some(v) = hashes.get(algo) {
            if !v.is_empty() {
                return (algo.to_lowercase(), v.clone());
            }
        }
    }
    ("md5".to_string(), package.attribute("hashMD5").unwrap_or("").to_string())
}

fn parse_dell_driverpack_date(value: &str) -> Option<Date> {
    // Dell's driver-pack catalog dateTime attribute is either a full
    // "YYYY-MM-DDTHH:MM:SS" timestamp or a bare "YYYY-MM-DD" date — no
    // timezone published either way, so this only ever extracts the date
    // part. Parsed manually (rather than via time's format-description
    // combinators) to keep this a simple, dependency-free string split,
    // mirroring the Python version's `datetime.strptime` fallback chain.
    if value.is_empty() {
        return None;
    }
    let date_part = value.get(0..10)?;
    let mut segments = date_part.split('-');
    let year: i32 = segments.next()?.parse().ok()?;
    let month_num: u8 = segments.next()?.parse().ok()?;
    let day: u8 = segments.next()?.parse().ok()?;
    let month = time::Month::try_from(month_num).ok()?;
    time::Date::from_calendar_date(year, month, day).ok()
}

fn make_tempdir() -> Result<PathBuf, String> {
    let base = std::env::temp_dir();
    let unique = format!(
        "waypoint-oem-{}-{}",
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0)
    );
    let dir = base.join(unique);
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    Ok(dir)
}
